### ISSUE TITLE
Error setting targetKey cookie in some cases

### ISSUE THREAD

**Issue by ghost** [plone/plone.app.drafts#13]
Plone 5.2.8 and plone.app.drafts 1.1.3 (installed with Mosaic)

In some cases, @@z3cform_validate_field will throw an error when attempting to process the cookies set by plone.app.drafts. I have noticed this when there is a space in the "path" value; this can be recreated by simply adding a new "News Item" which will try to set a cookie with path value like "/myplone/my-path/++add++News%s20Item". The problem is actually in the _setting the cookie_ step. https://github.com/plone/plone.app.drafts/blob/1.1.3/plone/app/drafts/current.py#L111 ZPublisher.cookie.path_converter will raise a ValueError because it finds a "%" in the path already which it interprets as meaning the path is already quoted. This is in Zope-4.8.1 (version pinned with Plone 5.2.8). At this point you have a busted cookie saved - ZPublisher has saved it with an empty dict {} in anticipation of filling it out in the next step, but this step has failed. @@z3cform_validate_field then fails trying to read this invalid cookie.

Since ZPublisher is already set up to convert urls for us, my assumption is that the simplest approach is to unquote the path and store the cookie that way. Before https://github.com/plone/plone.app.drafts/blob/1.1.3/plone/app/drafts/current.py#L111

```
from urllib.parse import unquote
path = unquote(path)
```

It may be that it is better to fix the value of path_safe in Zope-4.8.1. The regex ZPublisher.cookie.path_safe does not seem to account for + and - being safe in the path. (Shouldn't they be?)

edit: On second thought this does seem more likely to be a problem with ZPublisher as this function does not look correct. I will leave this here as a reference because it does affect plone.app.drafts

```
path_safe = compile(r"[a-zA-Z0-9_/%+-]*$").match

def path_converter(value):
    """convert *value* to a path.

    The convertion is based on ``absolute_url_path``.
    If *value* lacks this method, it is assumed to be a string.
    If the string contains ``%``, it is assumed that it is already
    quoted.
    """
    ap = getattr(aq_base(value), "absolute_url_path", None)
    if ap is not None:
        return value.absolute_url_path()
    if path_safe(value):
        return value
    if "%" in value:
        raise ValueError("not path safe but apparently quoted")
    return quote(value)
```
It fails the path_safe regex not because of the "%" in the path, but because of the plus/minus characters. This leads to the misleading ValueError raised. If the path either did not have the % or it did not have -/+ it would have converted.

**Comment by wohnlice** [comment]
It may be that it is better to fix the value of path_safe in Zope-4.8.1. The regex ZPublisher.cookie.path_safe does not seem to account for + and - being safe in the path. (Shouldn't they be?)

**Comment by ewohnlich** [comment]
This is addressed in https://github.com/zopefoundation/Zope/issues/1052
